File: sem1/hw7/7.3/tree.cpp
#include "tree.h"
#include <stdio.h>

int searchTokenStart(char *str, int i) {
    int sum = 0;
    if (str[i] == '(') {
        sum++;
        i++;
        while (sum != 0) {
            if (str[i] == '(')
                sum++;
            else if (str[i] == ')')
                sum--;
            i++;
        }
        while (str[i] == ' ') {
            i++;
        }
    } else {
        i++;
        while (str[i] != ' ') {
            i++;
        }
        while (str[i] == ' ') {
            i++;
        }
    }
    return i;
}

Tree *createTree() {
    return new Tree{nullptr};
}

void push(Node *&node, char *str, int i) {
    if (str[i] == '(') {
        node = new Node;
        i++;
        while (str[i] == ' ') {
            i++;
        }
        node->symbol = str[i];
        node->number = 0;
        node->right = nullptr;
        node->left = nullptr;
        i++;
        while (str[i] == ' ') {
            i++;
        }
        int firstTokenStart = i;
        int secondTokenStart = searchTokenStart(str, firstTokenStart);
        push(node->left, str, firstTokenStart);
        push(node->right, str, secondTokenStart);
        return;
    } else {
        node = new Node;
        node->number = 0;
        node->symbol = 'b';
        while ((str[i] >= '0') && (str[i] <= '9')) {
            node->number = node->number * 10 + str[i] - '0';
            i++;
        }
        node->left = nullptr;
        node->right = nullptr;
        return;
    }
}

void push(Tree *&tree, char *string, int i) {
    push(tree->root, string, i);
}

void print(Node *node) {
    if (node == nullptr)
        return;
    if (node->left != nullptr)
        printf("(");
    print(node->left);
    if (node->symbol != 'b')
        printf(" %c ", node->symbol);
    else
        printf("%d", node->number);
    print(node->right);
    if (node->left != nullptr) {
        printf(")");
    }

}

void print(Tree *tree) {
    print(tree->root);
}

int count(Node *tree) {
    if (tree->left == nullptr) {
        return tree->number;
    }

    switch (tree->symbol) {
        case '+':
            return count(tree->left) + count(tree->right);
        case '-':
            return count(tree->left) - count(tree->right);
        case '*':
            return count(tree->left) * count(tree->right);
        case '/':
            return count(tree->left) / count(tree->right);
    }
}

int count(Tree *tree) {
    return count(tree->root);
}

void deleteNode(Node *node) {
    if (node) {
        deleteNode(node->right);
        deleteNode(node->left);
        delete node;
    }

}

void deleteTree(Tree *tree) {
    deleteNode(tree->root);
    delete tree;
}

File: sem2/hw3/task3_2/ru/liamin/vladimir/Tree.java
package ru.liamin.vladimir;

/** Class implements tree */
public class Tree {

    private Node head;

    /**
     * Add elements to a tree
     * @param expression prefix arithmetic expression
     */
    public void push(char[] expression) {
        head = new Node();
        head.buildTree(expression, 0);
    }

    /** Print tree on console */
    public void print() {
        head.print();
    }

    /**
     * Count expression
     * @return expression value
     */
    public int count() {
        return head.count();
    }

    private class Node {
        private int number;
        private Node right;
        private Node left;
        private char symbol;

        private int searchTokenStart(char[] expression, int i) {
            int sum = 0;
            if (expression[i] == '(') {
                sum++;
                i++;
                while (sum != 0) {
                    if (expression[i] == '(')
                        sum++;
                    else if (expression[i] == ')')
                        sum--;
                    i++;
                }
                while (expression[i] == ' ') {
                    i++;
                }
            } else {
                i++;
                while (expression[i] != ' ') {
                    i++;
                }
                while (expression[i] == ' ') {
                    i++;
                }
            }
            return i;
        }

        private void buildTree(char[] expression, int i) {
            if (expression[i] == '(') {
                i++;
                while (expression[i] == ' ') {
                    i++;
                }
                right = null;
                left = null;
                number = 0;
                symbol = expression[i];
                i++;
                while (expression[i] == ' ') {
                    i++;
                }
                int firstTokenStart = i;
                int secondTokenStart = searchTokenStart(expression, firstTokenStart);
                left = new Node();

                left.buildTree(expression, firstTokenStart);
                right = new Node();
                right.buildTree(expression, secondTokenStart);
                return;
            } else {
                int value = 0;
                while ((expression[i] >= '0') && (expression[i] <= '9')) {
                    value = value * 10 + expression[i] - '0';
                    i++;
                }
                right = null;
                left = null;
                number = value;
                symbol = 'b';
                return;
            }
        }

        private void print() {
            if (left != null)
                System.out.print("(");
            if (left != null)
                left.print();

            if (symbol != 'b')
                System.out.print(" " + symbol + " ");
            else
                System.out.print(number);
            if (right != null)
                right.print();
            if (left != null) {
                System.out.print(")");
            }
        }

        private int count() throws ArithmeticException {
            if (left == null) {
                return number;
            }

            switch (symbol) {
                case '+':
                    return left.count() + right.count();
                case '-':
                    return left.count() - right.count();
                case '*':
                    return left.count() * right.count();
                case '/':
                    if (right.count() == 0)
                        throw new ArithmeticException("Division by zero");
                    return left.count() / right.count();
            }
            return number;
        }
    }
}
